sys: onoff: move transition function specification to separate structure

nordic-krch opened the pull request:
Extracted transition functions from onoff structure to external one which allows to keep them in flash.

Added ONOFF_SERVICE_DEFINE macro for static initialization.


Signed-off-by: Krzysztof Chruscinski <krzysztof.chruscinski@nordicsemi.no>

pabigot reviewed (commented):
I'd like to better understand why this is necessary.

Normally one would use `CONTAINER_OF` to determine the context relevant to the specific service:

```
struct clk_service {
  struct onoff_service srv;
  ... // context values here
};

static void clk_start(struct onoff_service *srv,
                      onoff_service_notify_fn notify);
{
   struct clk_service *ctx = CONTAINER_OF(srv, struct clk_service, srv);
   // ...
}
```

Alternatively, check which service structure is being passed:

```
static void clk_start(struct onoff_service *srv,
                      onoff_service_notify_fn notify);
{
   struct clk_context *ctx = NULL;
   if (srv == &hfclk_srv) {
     ctx = &clk_context[HFCLK_IDX];
   } else if (srv == &lfclk_srv) {
   } // ...
}
```

Why are those not satisfactory solutions?

nordic-krch commented:
@pabigot `CONTAINER_OF` is failing when things are splitted into ROM, RAM part. In my case, I keep constant things in ROM (start, stop task addresses) and i need to get them in transition function. Onoff service is stored in RAM.

pabigot reviewed (changes requested):
OK, I think the motivation for this for a one-off case of the clock service isn't very compelling, however it becomes more reasonable in the case of power rail management.  So I can accept this in principle, with some change requests.  Please do these, then I'll come back and look at the test infrastructure.

pabigot commented:
Thinking about it more: is it not true that, as long as `struct onoff_service` is stored in RAM, `CONTAINER_OF` allows the transition functions to access instance-specific information?  `struct onoff_service` can't be in ROM because it holds the client list as well as other state.

So I'm back to thinking this isn't worth doing.  #22420 might be, though.

nordic-krch committed:
sys: onoff: Move transition functions out of service struct

Extracted transition functions from onoff structure to external one
which allows to keep them in flash.
Added ONOFF_SERVICE_DEFINE macro for static initialization.

Signed-off-by: Krzysztof Chruscinski <krzysztof.chruscinski@nordicsemi.no>

nordic-krch commented:
@pabigot 
> struct onoff_service can't be in ROM because it holds the client list as well as other state.

yes, but it can be moved out into control block, then onff service is in rom with pointers to transition functions and control block.
Something like:
https://github.com/nordic-krch/zephyr/commit/d4b8f0ac32abd1bac418aac611bb8c3ac6bf8b28

nordic-krch commented:
I've added a commit with extracting transition functions into separate structure.

pabigot commented:
> > struct onoff_service can't be in ROM because it holds the client list as well as other state.
> 
> yes, but it can be moved out into control block, then onff service is in rom with pointers to transition functions and control block.

Why is that better than having `struct onoff_service` contain the RAM part, with a pointer to a function table which can (but need not be) in ROM?

I think we need to step back and try to use the service as originally designed, rather than jump right into changing it, given that no concern that's been raised prevents it from being used in the identified cases.

nordic-krch commented:
> Why is that better than having struct onoff_service contain the RAM part, with a pointer to a function table which can (but need not be) in ROM?

another 4 bytes moved to ROM in case context and transition functions are static (which might be most of the cases)

pabigot commented:
> > Why is that better than having struct onoff_service contain the RAM part, with a pointer to a function table which can (but need not be) in ROM?
> 
> another 4 bytes moved to ROM in case context and transition functions are static (which might be most of the cases)

Yes, but that cost is decreased by the additional code required to pass another argument that could otherwise be deduced from an existing argument.

Given the stage of the release cycle, and my hope that we could fix the clock startup code for 2.2, I would really like to see a solution to clock startup that uses the service API as it currently exists.  Then we can see how well it worked and consider what changes might need to be made, taking into account other use cases as well.

nordic-krch commented:
what about extracting transition functions to separate struct? I agree that it is an enhancement but on the other hand it impacts api (init function) thus it would be good to get that to 2.2.

nordic-krch commented:
regarding adding user_data, i agree that you can always manipulate your data to be able to get container and get pointer you want. E.g. if you want device pointer and onoff struct is somewhere in driver data you can add:
```
struct driver_data {
  struct device *dev; //reference to device needed by onoff service
  struct onoff_service service;
}
```
Drawback of that solution is that it's more implicit than passing dev as context to the service. Advantage of that is that onoff service does not have additional field which might not always be need.

pabigot commented:
> what about extracting transition functions to separate struct? I agree that it is an enhancement but on the other hand it impacts api (init function) thus it would be good to get that to 2.2.

Yes, I think that's OK, and is consistent with how driver APIs are normally managed.  I can imagine only a few cases where being able to change the transition functions at runtime could be useful, and then it could be stored in RAM.  Please proceed with a proposal for #22420.

> regarding adding context, i agree that you can always manipulate your data to be able to get container and get pointer you want. E.g. if you want device pointer and onoff struct is somewhere in driver data you can add:
> 
> ```
> struct driver_data {
>   struct device *dev; //reference to device needed by onoff service
>   struct onoff_service service;
> }
> ```
> 
> Drawback of that solution is that it's more implicit than passing dev as context to the service. Advantage of that is that onoff service does not have additional field which might not always be need.

Yes, and it's consistent with most `CONTAINER_OF` solutions currently in Zephyr.  So we'll hold off on this API change?

nordic-krch commented:
Removed context parameter and kept extraction of transition functions into separate struct.

pabigot commented on include/sys/onoff.h:
```diff
@@ -102,6 +108,29 @@ typedef void (*onoff_service_notify_fn)(struct onoff_service *srv,
+	 * @ref ONOFF_SERVICE_RESET_SLEEPS as appropriate in flags.
+	 */
+	onoff_service_transition_fn reset;
+};
```
These should be documentation comments.  If they are multi-line, they should follow the implicit-brief style used for all other comments in this file, e.g.

```
	/**
	 * The function used to (initiate a) transition from off to on.
	 *
	 * This must not be null.
	 */
```

The information about `flags` doesn't really belong here if `flags` isn't part of this structure.  Thinking about it, though, the set of flags (restricted to the sleeps ones) *should* be stored here because it's tightly bound to the specific functions that are stored here.  Then the field here should be used to initialize the RAM copy in the service structure (which is in RAM because in principle I want services to be able to use the extra bits to record their own state).

So please add a `u16_t flags` field and update the initializer macros to take it.  Then the initializer macro for the service structure does not need to incorporate flags.

Hope that's clear.

nordic-krch replied:
Agree that flags belong here but then it cannot be stored in ROM. Maybe flags should be split as they are combined into one variable mainly for optimization reasons.

pabigot commented on include/sys/onoff.h:
```diff
@@ -141,14 +162,44 @@ struct onoff_service {
 	u16_t refs;
 };
 
+#define ONOFF_SERIVCE_TRANSITIONS_INITIALIZER(_start, _stop, _reset) {  \
```
add `flags` per above

pabigot commented on include/sys/onoff.h:
```diff
@@ -141,14 +162,44 @@ struct onoff_service {
-		.stop = _stop,					   \
-		.reset = _reset,				   \
-		.flags = _flags,				   \
+#define ONOFF_SERVICE_INITIALIZER(_transitions, _flags) { \
```
remove `_flags` per above, initialize `.flags = (_transitions)->flags,`

pabigot commented on include/sys/onoff.h:
```diff
@@ -141,14 +162,44 @@ struct onoff_service {
+ * @param _reset a function used to clear errors and force the service to an off
+ * state. Can be null.
+ *
+ * @param _flags any or all of the flags mentioned above,
```
remove `_flags`

pabigot commented on tests/lib/onoff/src/main.c:
```diff
@@ -167,26 +167,36 @@ static void test_service_init_validation(void)
 	int rc;
 	struct onoff_service srv;
+	const struct onoff_service_transitions null_transitions =
+		ONOFF_SERIVCE_TRANSITIONS_INITIALIZER(NULL, NULL, NULL);
```
Typo in name `SERIVCE`.

pabigot reviewed (changes requested):
I think this is going to be a good enhancement.  Please address these first then I'll take another pass.

pabigot commented on include/sys/onoff.h:
```diff
@@ -141,14 +172,55 @@ struct onoff_service {
+ *
+ * @param _flags any or all of the flags from enum onoff_service_flags.
+ */
+#define ONOFF_SERIVCE_TRANSITIONS_INITIALIZER(_start, _stop, _reset, _flags) { \
```
Here's where the misspelling needs to be fixed.

nordic-krch commented on include/sys/onoff.h:
```diff
@@ -49,6 +49,12 @@ enum onoff_service_flags {
 	 */
 	ONOFF_SERVICE_RESET_SLEEPS      = BIT(2),
 
+	ONOFF_SERVICE_FLAGS_LAST        = ONOFF_SERVICE_RESET_SLEEPS,
```
won't it be cleaner for the user which flags are for configuration?

pabigot replied:
Possibly, but there's a bigger risk of creating overlaps if they aren't defined contiguously.  Also a user that wants to leverage available flags needs to be aware of which bits are in use.

nordic-krch commented on lib/os/onoff.c:
```diff
@@ -11,11 +11,6 @@
 #define CLIENT_NOTIFY_METHOD_MASK 0x03
 #define CLIENT_VALID_FLAGS_MASK 0x07
 
-#define SERVICE_CONFIG_FLAGS	     \
```
i used in for same validation during static initialization:
```
BUILD_ASSERT_MSG((_flags & ~ONOFF_SERVICE_FLAGS_MASK) == 0,	   \
				"Flags misused");			   \
```
that's why it was moved to the header initially and then converted to MASK.

pabigot replied:
For now I'd prefer to expand it out in this location:
```
BUILD_ASSERT_MSG(((_flags) & (ONOFF_SERVICE_START_SLEEPS
                              | ONOFF_SERVICE_STOP_SLEEPS
                              | ONOFF_SERVICE_RESET_SLEEPS)) == (_flags),
                 "Non-transition flags specified");
```
The current name for the mask isn't accurate; here we're checking that only transition-function--specific flags are provided.  Also doing it here is more clear to anybody who wants to know what's allowed.

pabigot reviewed (changes requested):
A few more tweaks.

pabigot commented:
@nordic-krch Would it be OK if I adopted this commit into #22853 so we don't have to deal with conflicts based on which gets merged first?

nordic-krch commented:
@pabigot sure

pabigot commented:
This is now incorporated into #22853.  DNM until it's merged or rejected there.
